### PR TITLE
Cleanup Parallel Backend

### LINKED ISSUE #360: Add options to ray.remote in MapReduce and add num_cpus, num_gpus explicitly clearification.
We should add `ray.remote` options to our MapReduce implementation as well as make number of CPU's and GPU's explicitly. This helps to ensure clarity on how much resources are used and adds robustness if ray changes the default behaviour in the future. There are two affected code parts:

1. https://github.com/appliedAI-Initiative/pyDVL/blob/d5381d7a51e26a7e8d904c91ac0f01d67717bb9e/src/pydvl/utils/parallel/map_reduce.py#L193
2. https://github.com/appliedAI-Initiative/pyDVL/blob/d5381d7a51e26a7e8d904c91ac0f01d67717bb9e/src/pydvl/utils/parallel/map_reduce.py#L215

and a possible solution would look like:

```python
fn = self._wrap_function(self._fn, num_cpus=self.num_cpus_per_task, num_gpus=self.num_gpus_per_task, **options)
```

### PR DESCRIPTION
**Pull request by AnesBenmerzoug**
### Description

This PR closes #360 #535 #536 and addresses part of #385 

### Changes

- Adapted docstrings and code to rely on user to configure joblib and initialize ray with desired parameters.
- Set minimum version of joblib to 1.3.0 in order to support `parallel_config` context manager.
- Deprecated most fields of `ParallelConfig` except for `backend` in preparation for removing it altogether in favour of letting the user instantiate the parallel backend instance and passing it to the method.
- Moved parallelization and caching documentation to separated advanced usage page and expanded it.
- Fixed caching and dask influence calculator docstrings.

### Checklist

- [ ] Wrote Unit tests (if necessary)
- [X] Updated Documentation (if necessary)
- [ ] Updated Changelog
- [ ] ~If notebooks were added/changed, added boilerplate cells are tagged with `"tags": ["hide"]` or `"tags": ["hide-input"]`~

### CODE REVIEW

**Review by mdbenito — COMMENTED** [review]
Cool. Simple is always better 👍🏽 I left a few minor comments and a few questions. Maybe we can talk on Monday about how the final usage pattern will look like?

File: tests/value/shapley/test_montecarlo.py [hunk 1]
```diff
@@ -134,7 +134,11 @@ def test_seed(
...
 @pytest.mark.parametrize("num_samples, delta, eps", [(6, 0.1, 0.1)])
 @pytest.mark.parametrize(
-    "fun", [ShapleyMode.PermutationMontecarlo, ShapleyMode.CombinatorialMontecarlo]
+    "fun",
+    [
+        ShapleyMode.PermutationMontecarlo,
+        pytest.param(ShapleyMode.CombinatorialMontecarlo, marks=pytest.mark.skip),
+    ],
 )
 def test_hoeffding_bound_montecarlo(
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply tests/value/shapley/test_montecarlo.py:143]: Alright. I will mark it as skipped for now and we can document it later.
We could actually move the tolerate fixture/plugin to a separate repository and open-source. I am sure someone else will find it useful.

File: tests/utils/test_caching.py [hunk 2]
```diff
@@ -19,9 +19,16 @@
...
 from pydvl.utils.types import Seed
 
+from ..conftest import num_workers
+
 logger = logging.getLogger(__name__)
 
 
+@pytest.fixture(scope="module")
+def parallel_config():
+    return ParallelConfig(backend="joblib", n_cpus_local=num_workers())
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply tests/utils/test_caching.py:29]: It's unfortunately deprecated but not yet replaced. I am planning to do that in a separate PR. I could also undeprecate it here and do that in the next PR but I was too lazy to undo it.

File: src/pydvl/parallel/backends/joblib.py [hunk 3]
```diff
@@ -25,14 +25,29 @@ class JoblibParallelBackend(BaseParallelBackend, backend_name="joblib"):
...
+        parallel_backend = init_parallel_backend(config)
+        ```
+
+    ??? Example
+        ``` python
+        import joblib
+        from pydvl.parallel import init_paralle_backend, ParallelConfig
+        with joblib.parallel_config(verbose=100):
+            config = ParallelConfig(backend="joblib")
+            parallel_backend = init_parallel_backend(config)
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/parallel/backends/joblib.py:41]: The suggested pattern is very similar to [sklearn's parallelization pattern](https://scikit-learn.org/stable/computing/parallelism.html), except for the parallel backend thing, so I would think that users would be familiar with it.

I preferred this pattern over hiding joblib completely away in our class because we would have to duplicate its entire configuration interface and keep passing the values around. I didn't like it from a maintainability point of view.

You're right about the private constructors. I don't remember why we needed them as well. I was going to remove them in a follow-up PR anyway.

File: tests/value/shapley/test_montecarlo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply tests/value/shapley/test_montecarlo.py:147]: If you are up to it... It's a non-negligible amount of effort.

File: src/pydvl/parallel/backends/joblib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/parallel/backends/joblib.py:41]: I guess if the ParallelConfig line goes away, it's only two lines, which is better. But still, having our init and their init is very weird. Can we do our init from theirs?
```python
with joblib.parallel_config(blah) as pc:
    pb = JoblibParallelBackend(pc)
    compute_schtuff(u, params, parallel_backend=pb)
```
Because the discussion about just using joblib as our backend and using it to talk to dask/ray/whatever ended up showing that we couldn't, correct?

[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/parallel/backends/joblib.py:41]: With the change I want to do in the follow-up PR this is how it would look like:

```python
with joblib.parallel_config(blah):
    pb = JoblibParallelBackend()
    compute_schtuff(u, params, parallel_backend=pb)
```

We only use `joblib` inside `MapReduceJob` and the change in this PR makes it so that the users don't have to pass joblib parameters to our classes and instead rely on joblib's mechanisms to do that.

The idea of just using joblib for parallelization didn't go through because we have the futures interface and that one doesn't use joblib.

File: tests/value/shapley/test_montecarlo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply tests/value/shapley/test_montecarlo.py:147]: I actually thought about doing it a long time ago but didn't bother till now. The way it is written should make it straightforward to extract into a separate repository. I will create an issue for it.

[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply tests/value/shapley/test_montecarlo.py:147]: I created issue #539 to track that.

File: src/pydvl/parallel/backends/joblib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/parallel/backends/joblib.py:41]: I see, ok. That is fine, agreed. One line config, one line init backend, one line use.

About just joblib: I remember now, thanks.

And to the long-running discussion: I still think that we should ditch futures in favour of workers+queues. There's no good way to interrupt otherwise, or to send information back to samplers (e.g. for adaptive stratified sampling). Both are requirements of several algorithms.

Plus with workers we reuse processes more efficiently and probably transfer less data (or read/copy less often from local storage / ram). And with a decent job/worker object with a couple of hooks built in, algorithm code can remain as simple as it is now. Just subclass and fill in the gaps with your function, and you're good to go.

[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/parallel/backends/joblib.py:41]: I am not against the idea of using queues, I am just worried that it would bring a new type of bugs that we have to worry about. We should do it properly and take inspiration from other projects. I like for example [Optuna](https://github.com/optuna/optuna/tree/master)'s parallelization approach that uses a database to completely decouple the main process from the workers. They also have samplers that even though are not the same as ours, can still give us an idea of how to implement them.

For now though, I think we should just clean things up for the next release and for pydata.

[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/parallel/backends/joblib.py:41]: I haven't looked at optuna, but I would assume they use a database for an altogether different reason than decoupling. Namely, keeping track of hyperparameter choices and experiment results. We don't want to keep track of these things, or if we do, they are associated to a result set, which we don't persist. Introducing a big external dependency which requires deployment of a service (and thus might be problematic in many environments) just to avoid keeping track of a couple of queues seems overkill to me. I hope we can find a better, and simpler solution

File: docs/getting-started/first-steps.md [hunk 4]
```diff
@@ -11,7 +11,7 @@ alias:
     Make sure you have read [[getting-started#installation]] before using the library. 
     In particular read about which extra dependencies you may need.
 
-## Main concepts
+## Main Concepts
```
Comment (mdbenito) [inline docs/getting-started/first-steps.md:14]: Why this change? We don't capitalize titles elsewhere, do we?

File: src/pydvl/parallel/config.py [hunk 5]
```diff
@@ -1,21 +1,22 @@
-import logging
+import warnings
 from dataclasses import dataclass
 from typing import Literal, Optional, Tuple, Union
 
 __all__ = ["ParallelConfig"]
 
 
+# TODO: delete this class once it's made redundant in v0.10.0
```
Comment (mdbenito) [inline src/pydvl/parallel/config.py:8]: won't this get lost? Elsewhere I used
```python
# This string for the benefit of deprecation searches:
# remove_in="0.9.0"
```
since that is the format of the deprecated lib

**Review by mdbenito — DISMISSED** [review]
Looking good 👍🏽

File: src/pydvl/parallel/backends/joblib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/parallel/backends/joblib.py:41]: They use the database for both purposes. In the case of queues, we also need to deploy an external dependency.
Anyway, I will prepare a prototype or at least a design document before mentioning that again.

File: docs/getting-started/first-steps.md (hunk 4 above)
Comment (AnesBenmerzoug) [reply docs/getting-started/first-steps.md:14]: We actually do in quite a lot of places. The inconsistency highly depends on who wrote the documentation. I will make everything consistent.

File: src/pydvl/parallel/config.py (hunk 5 above)
Comment (AnesBenmerzoug) [reply src/pydvl/parallel/config.py:8]: I assumed that it however is doing this would search for a `TODO` I will add your suggestion as well.

**Review by mdbenito — APPROVED** [review]
